File: utils/gate.py
"""
PredictionGate — three-layer temporal filter applied after every inference call.

Layer 1: Majority vote over a sliding window of the last N frames.
Layer 2: Hold queue — the voted label must be stable for M consecutive frames.
Layer 3: Cooldown — a minimum gap of K frames between two accepted predictions.

This is the canonical Python implementation; the browser TypeScript version in
``browser/gate.ts`` must mirror this logic exactly.

Requirements: 2.4, 9.5, 9.6
"""

from __future__ import annotations

from collections import deque
from typing import Optional


class PredictionGate:
    """Three-layer temporal filter for stable sign commitment.

    Args:
        vote_window:    Number of recent frames used for majority voting (default 7).
        hold_frames:    Number of consecutive frames the voted label must be
                        stable before it is accepted (default 12).
        confidence_threshold: Minimum per-frame confidence required for a frame
                        to contribute to the vote window (default 0.8).
        cooldown_frames: Number of frames to suppress new predictions after an
                        acceptance event (default 20).
    """

    def __init__(
        self,
        vote_window: int = 7,
        hold_frames: int = 12,
        confidence_threshold: float = 0.8,
        cooldown_frames: int = 20,
    ) -> None:
        if vote_window < 1:
            raise ValueError("vote_window must be >= 1")
        if hold_frames < 1:
            raise ValueError("hold_frames must be >= 1")
        if not (0.0 <= confidence_threshold <= 1.0):
            raise ValueError("confidence_threshold must be in [0, 1]")
        if cooldown_frames < 0:
            raise ValueError("cooldown_frames must be >= 0")

        self.vote_window = vote_window
        self.hold_frames = hold_frames
        self.confidence_threshold = confidence_threshold
        self.cooldown_frames = cooldown_frames

        # Internal state
        self._window: deque[str] = deque(maxlen=vote_window)
        self._hold_label: Optional[str] = None
        self._hold_count: int = 0
        self._cooldown_remaining: int = 0

    # ------------------------------------------------------------------
    # Public API
    # ------------------------------------------------------------------

    def update(self, label: str, confidence: float) -> Optional[str]:
        """Feed one frame's prediction into the gate.

        Args:
            label:      Predicted class label for this frame.
            confidence: Softmax probability in [0.0, 1.0] for the predicted label.

        Returns:
            The accepted label string if the gate fires, or ``None`` if no
            commitment is made this frame.
        """
        # Decrement cooldown counter
        if self._cooldown_remaining > 0:
            self._cooldown_remaining -= 1
            # Still suppress: push to window but do not accept
            if confidence >= self.confidence_threshold:
                self._window.append(label)
            else:
                self._window.append("")  # low-confidence frame counts as no vote
            # Reset hold state during cooldown
            self._hold_label = None
            self._hold_count = 0
            return None

        # Layer 1 — majority vote
        if confidence >= self.confidence_threshold:
            self._window.append(label)
        else:
            self._window.append("")  # low-confidence frame: blank vote

        voted_label = self._majority_vote()

        if voted_label is None:
            # No majority — reset hold
            self._hold_label = None
            self._hold_count = 0
            return None

        # Layer 2 — hold queue
        if voted_label == self._hold_label:
            self._hold_count += 1
        else:
            self._hold_label = voted_label
            self._hold_count = 1

        if self._hold_count >= self.hold_frames:
            # Layer 3 — commit and enter cooldown
            accepted = self._hold_label
            self._hold_label = None
            self._hold_count = 0
            self._window.clear()
            self._cooldown_remaining = self.cooldown_frames
            return accepted

        return None

    def reset(self) -> None:
        """Clear all internal state (call when the user clears the buffer or
        the hand is lost)."""
        self._window.clear()
        self._hold_label = None
        self._hold_count = 0
        self._cooldown_remaining = 0

    # ------------------------------------------------------------------
    # Internal helpers
    # ------------------------------------------------------------------

    def _majority_vote(self) -> Optional[str]:
        """Return the label with a strict majority (> 50 %) in the window,
        or None if no such label exists.  Blank votes ("") never win."""
        if not self._window:
            return None

        counts: dict[str, int] = {}
        for lbl in self._window:
            if lbl:  # ignore blank votes
                counts[lbl] = counts.get(lbl, 0) + 1

        if not counts:
            return None

        best_label = max(counts, key=lambda k: counts[k])
        if counts[best_label] > len(self._window) / 2:
            return best_label
        return None

File: browser/gate.ts
/**
 * PredictionGate — three-layer temporal filter applied after every inference call.
 *
 * Layer 1: Majority vote over a sliding window of the last N frames.
 * Layer 2: Hold queue — the voted label must be stable for M consecutive frames.
 * Layer 3: Cooldown — a minimum gap of K frames between two accepted predictions.
 *
 * This is a direct TypeScript port of the canonical Python implementation in
 * ``utils/gate.py`` — the logic is identical.
 *
 * Requirements: 2.4, 9.5, 9.6
 */

/**
 * Three-layer temporal filter for stable sign commitment.
 *
 * @param voteWindow      Number of recent frames used for majority voting (default 7).
 * @param holdFrames      Number of consecutive frames the voted label must be
 *                        stable before it is accepted (default 12).
 * @param cooldownFrames  Number of frames to suppress new predictions after an
 *                        acceptance event (default 20).
 * @param confidenceThreshold  Minimum per-frame confidence required for a frame
 *                        to contribute to the vote window (default 0.8).
 */
export class PredictionGate {
  readonly voteWindow: number;
  readonly holdFrames: number;
  readonly cooldownFrames: number;
  readonly confidenceThreshold: number;

  // Internal state — mirroring Python's deque / instance variables
  private _window: string[];          // circular-ish buffer capped at voteWindow
  private _holdLabel: string | null;
  private _holdCount: number;
  private _cooldownRemaining: number;

  constructor(
    voteWindow = 7,
    holdFrames = 12,
    cooldownFrames = 20,
    confidenceThreshold = 0.8,
  ) {
    if (voteWindow < 1) {
      throw new RangeError("voteWindow must be >= 1");
    }
    if (holdFrames < 1) {
      throw new RangeError("holdFrames must be >= 1");
    }
    if (confidenceThreshold < 0.0 || confidenceThreshold > 1.0) {
      throw new RangeError("confidenceThreshold must be in [0, 1]");
    }
    if (cooldownFrames < 0) {
      throw new RangeError("cooldownFrames must be >= 0");
    }

    this.voteWindow = voteWindow;
    this.holdFrames = holdFrames;
    this.cooldownFrames = cooldownFrames;
    this.confidenceThreshold = confidenceThreshold;

    this._window = [];
    this._holdLabel = null;
    this._holdCount = 0;
    this._cooldownRemaining = 0;
  }

  // ---------------------------------------------------------------------------
  // Public API
  // ---------------------------------------------------------------------------

  /**
   * Feed one frame's prediction into the gate.
   *
   * @param label       Predicted class label for this frame.
   * @param confidence  Softmax probability in [0.0, 1.0] for the predicted label.
   * @returns The accepted label string if the gate fires, or null if no
   *          commitment is made this frame.
   */
  update(label: string, confidence: number): string | null {
    // Decrement cooldown counter
    if (this._cooldownRemaining > 0) {
      this._cooldownRemaining -= 1;
      // Still suppress: push to window but do not accept
      this._windowAppend(confidence >= this.confidenceThreshold ? label : "");
      // Reset hold state during cooldown
      this._holdLabel = null;
      this._holdCount = 0;
      return null;
    }

    // Layer 1 — majority vote
    this._windowAppend(confidence >= this.confidenceThreshold ? label : "");

    const votedLabel = this._majorityVote();

    if (votedLabel === null) {
      // No majority — reset hold
      this._holdLabel = null;
      this._holdCount = 0;
      return null;
    }

    // Layer 2 — hold queue
    if (votedLabel === this._holdLabel) {
      this._holdCount += 1;
    } else {
      this._holdLabel = votedLabel;
      this._holdCount = 1;
    }

    if (this._holdCount >= this.holdFrames) {
      // Layer 3 — commit and enter cooldown
      const accepted = this._holdLabel as string;
      this._holdLabel = null;
      this._holdCount = 0;
      this._window = [];
      this._cooldownRemaining = this.cooldownFrames;
      return accepted;
    }

    return null;
  }

  /**
   * Clear all internal state (call when the user clears the buffer or
   * the hand is lost).
   */
  reset(): void {
    this._window = [];
    this._holdLabel = null;
    this._holdCount = 0;
    this._cooldownRemaining = 0;
  }

  // ---------------------------------------------------------------------------
  // Internal helpers
  // ---------------------------------------------------------------------------

  /**
   * Append a label to the window, evicting the oldest entry when the window
   * is full — mirroring Python's `deque(maxlen=voteWindow)`.
   */
  private _windowAppend(label: string): void {
    this._window.push(label);
    if (this._window.length > this.voteWindow) {
      this._window.shift();
    }
  }

  /**
   * Return the label with a strict majority (> 50 %) in the window,
   * or null if no such label exists.  Blank votes ("") never win.
   */
  private _majorityVote(): string | null {
    if (this._window.length === 0) {
      return null;
    }

    const counts: Record<string, number> = {};
    for (const lbl of this._window) {
      if (lbl !== "") {   // ignore blank votes
        counts[lbl] = (counts[lbl] ?? 0) + 1;
      }
    }

    const labels = Object.keys(counts);
    if (labels.length === 0) {
      return null;
    }

    let bestLabel = labels[0];
    for (const lbl of labels) {
      if (counts[lbl] > counts[bestLabel]) {
        bestLabel = lbl;
      }
    }

    // Strict majority: count must exceed half the window size
    if (counts[bestLabel] > this._window.length / 2) {
      return bestLabel;
    }
    return null;
  }
}
